src/responder.ts

```diff
@@ -150,83 +176,329 @@ export class EthereumDedicatedResponder extends EthereumResponder {
+                const enoughConfirmationsPromise = waitForConfirmations(this.signer.provider, tx.hash, this.confirmationsRequired);
+
 
-                // Wait for enough confirmations before declaring success
-                const confirmationsPromise = waitForConfirmations(this.signer.provider, tx.hash, this.confirmationsRequired);
                 // ...but stop with error if no new blocks come for too long
+                // TODO: make sure this does not cause memory leaks
                 const noNewBlockPromise = new Promise((_, reject) => {
                     const intervalHandle = setInterval( () => {
                         // milliseconds since the last block was received (or the responder was instantiated)
                         const msSinceLastBlock = Date.now() - this.timeLastBlockReceived;
-                        if (msSinceLastBlock > 60*1000) {
+                        if (msSinceLastBlock > EthereumDedicatedResponder.WAIT_TIME_FOR_NEW_BLOCK) {
                             clearInterval(intervalHandle)
                             reject(new NoNewBlockError(`No new block was received for ${Math.round(msSinceLastBlock/1000)} seconds; provider might be down.`));
                         }
                     }, 1000);
                 });
 
-                await Promise.race([confirmationsPromise, noNewBlockPromise]);
+                // First, wait to get at least 1 confirmation, but throw an error if the transaction is stuck
+                // (that is, new blocks are coming, but the transaction is not included)
+                await Promise.race([
+                    firstConfirmationPromise,
+                    firstConfirmationTimeoutPromise,
+                    noNewBlockPromise
+                ]);
+
+                // Then, wait to get at enough confirmations; now only throw an error if there is a reorg
+                await Promise.race([
+                    enoughConfirmationsPromise,
+                    noNewBlockPromise
+                ]);
 
                 // The response has now enough confirmations to be considered safe.
-                this.responseFlow.status = ResponseState.Success;
-                this.asyncEmit(ResponderEvent.ResponseConfirmed, this.responseFlow);
+                responseFlow.state = ResponseState.Success;
+                this.asyncEmit(ResponderEvent.ResponseConfirmed, responseFlow);
 
                 this.destroy();
 
                 return;
             } catch (doh) {
-                this.asyncEmit(ResponderEvent.AttemptFailed, this.responseFlow, doh);
+                this.asyncEmit(ResponderEvent.AttemptFailed, responseFlow, doh);
+
+                if (doh instanceof StuckTransactionError) {
+                    // Double the gas price before the next attempt
+                    // TODO: think of better strategies (e.g.: check network conditions again)
+                    this.gasPrice = this.gasPrice.mul(2);
```

Comment:
I think this should be a class: GasPolicy with two methods, getStartGasPrice and getEscalateGasPrice(currentGasPrice). The second of these just multiplies price. You can then inject a gasPolicy into the responder via the constructor, and you can test the policy separately which will be useful when they get more complicated.

Reply:
Did something like this in 89f5650a3281d0c269ca2d0a4b2a297a7ffe928c, except I made the methods return a promise, to keep open the possibility of querying the provider (for example).

Reply:
As mentioned below, when you need to query the provider you should refactor the code at that point.